# Obteniendo metadata de las canciones utilizando Spotify APIs

## Definimos la función `get_info` que busca canciones en la base de datos de spotify a través del API `v1/search` utilizando nombre de la canción y album por default

### Al pasar el parámetro `strict=False`, busca solamente por artista. Esto solo se utiliza en el caso del conjunto de datos relativos a 2020, en donde se revisó una por una para ver que fueran las canciones correctas



In [1]:
import requests, json, logging
import pandas as pd

def get_info(song_name = '', artist_name = '', req_type = 'track',force=False):
    r = requests.post('https://accounts.spotify.com/api/token', headers = {'Authorization': 'Basic NDM0YmFiM2VhNmM2NDg2MmI3NmJkYWUwOTA0NmU2Njg6ZjFlZmFhZmM5MjA1NDFiYzkyZGNlMTk2MzBhZjk1NzE='}, data= {'grant_type': 'client_credentials'})
    token = 'Bearer {}'.format(r.json()['access_token'])
    headers = {'Authorization': token, "Accept": 'application/json', 'Content-Type': "application/json"}
    try:
        payload = {"q" : "artist:{} track:{}".format(artist_name, song_name), "type": req_type, "limit": "1"}
        res = requests.get('https://api.spotify.com/v1/search', params = payload, headers = headers)
        res = res.json()['tracks']['items'][0]
        year = res['album']['release_date'][:4]
        artist_id = res['artists'][0]['id']
        track_id = res['id']
        track_pop = res['popularity'] 
    except:
        print(f'Unable to get {song_name} from {artist_name}')
        if force:
            print(f'Trying {song_name} by only artist {artist_name}')
            payload = {"q" : "artist:{}".format(artist_name, song_name), "type": req_type, "limit": "1"}
            res = requests.get('https://api.spotify.com/v1/search', params = payload, headers = headers)
            res = res.json()['tracks']['items'][0]
            year = res['album']['release_date'][:4]
            artist_id = res['artists'][0]['id']
            track_id = res['id']
            track_pop = res['popularity'] 
        
            
        
    res = requests.get('https://api.spotify.com/v1/audio-analysis/{}'.format(track_id), headers = headers)
    res = res.json()['track']
    
    res2 = requests.get('https://api.spotify.com/v1/audio-features/{}'.format(track_id), headers = headers)
    res2 = res2.json()
    
    duration = res['duration']
    end_fade = res['end_of_fade_in']
    key = res['key']
    key_con = res['key_confidence']
    loud = res['loudness']
    mode = res['mode']
    mode_con = res['mode_confidence']
    start_fade = res['start_of_fade_out']
    temp = res['tempo']
    time_sig = res['time_signature']
    time_sig_con = res['time_signature_confidence']
    danceability=res2['danceability']
    energy=res2['energy']
    key=res2['key']
    loudness=res2['loudness']
    mode=res2['mode']
    speechiness=res2['speechiness']
    acousticness=res2['acousticness']
    instrumentalness=res2['instrumentalness']
    liveness=res2['liveness']
    valence=res2['valence']
    tempo=res2['tempo']
    duration_ms=res2['duration_ms']
    time_signature=res2['time_signature']
    
    res = requests.get('https://api.spotify.com/v1/artists/{}'.format(artist_id), headers = headers)
    artist_hot = res.json()['popularity']/100
    
    return {'duration': duration,
            'key': key,
            'loudness': loud,
            'mode': mode,
            'tempo': temp,
            'artist_hotttnesss': artist_hot,
            'end_of_fade_in': end_fade,
            'start_of_fade_out': start_fade,
            'mode_confidence': mode_con,
            'key_confidence': key_con,
            'time_signature': time_sig,
            'time_signature_confidence': time_sig_con,
            'year': year,
            'popularity': track_pop,
            'danceability':danceability,
            'energy':energy,
            'speechiness':speechiness,
            'acousticness':acousticness,
            'instrumentalness':instrumentalness,
            'liveness':liveness,
            'valence':valence,
            'tempo':tempo,
            'duration_ms':duration_ms,
           }

### Bucle para iterar sobre todos los archivos `20*.csv` es decir (2002.csv, 2003.csv...,2019.csv)

In [2]:
import glob
lista=[]
for filename in glob.iglob('20*.csv'):
    lista.append(pd.read_csv(filename))
    
todos=pd.concat(lista)
todos.reset_index(drop=True,inplace=True)
todos

Draw          Country                 Artist                   Song  \
0       1          Romania   Mădălina and Andrada       Salvați planeta!   
1       2          Armenia                 Monica       Im Ergi Hnchyune   
2       3          Belarus  Dasha, Alina & Karyna      Serdtse Belarusi    
3       4           Russia          Mihail Puntov            Spit angel    
4       5           Greece        Niki Yiannouchu         Kapoia nychta    
..    ...              ...                    ...                    ...   
271    12          Belgium                Lindsay              Mes rêves   
272    13            Malta       Thea and Friends         Make It Right!   
273    14           Norway                  Malin  Sommer og skolefri[a]   
274    15            Spain           Antonio José       Te traigo flores   
275    16  Belarus Belarus          Ksenia Sitnik             My vmeste    

                    Language  Place  Points Language(s)  
0                   Romanian    9.0    58.0         NaN  
1                   Armenian    8.0    59.0         NaN  
2    Russian, Belarusian[22]    6.0    86.0         NaN  
3                    Russian    7.0    73.0         NaN  
4                      Greek   14.0    19.0         NaN  
..                       ...    ...     ...         ...  
271                   French   10.0    63.0         NaN  
272                  English   16.0    18.0         NaN  
273                Norwegian    3.0   123.0         NaN  
274                  Spanish    2.0   146.0         NaN  
275                  Russian    1.0   149.0         NaN  

[276 rows x 8 columns]

### Obtenemos 276 filas, pero terminaremos con muchas menos ya que un conjunto pequeño reducido de canciones se encuentra dentro de la biblioteca de Spotify (lo cual es necesario para llamar a su API y obtener los datos que nos interesan)

### Después de llamar a los APIs de Spotify, terminamos con nuestro conjunto de entrenamient

In [3]:
lista=[]
for index, row in todos.iterrows():
    try:
        dic = get_info(song_name=row["Song"],artist_name=row["Artist"])
        dic['Points']=row['Points']
        dic['Country']=row['Country']
        lista.append(pd.DataFrame(dic,index=[index]))
        
    except Exception as e:
        print(e,f"missing {row['Song']}")

Unable to get Salvați planeta! from Mădălina and Andrada
local variable 'track_id' referenced before assignment missing Salvați planeta!
Unable to get Im Ergi Hnchyune from Monica
local variable 'track_id' referenced before assignment missing Im Ergi Hnchyune
Unable to get Serdtse Belarusi  from Dasha, Alina & Karyna
local variable 'track_id' referenced before assignment missing Serdtse Belarusi 
Unable to get Spit angel  from Mihail Puntov
local variable 'track_id' referenced before assignment missing Spit angel 
Unable to get Kapoia nychta  from Niki Yiannouchu
local variable 'track_id' referenced before assignment missing Kapoia nychta 
Unable to get Bzz.. from Bzikebi
local variable 'track_id' referenced before assignment missing Bzz..
'track' missing Shut Up
Unable to get Edna mechta  from Krastyana Krasteva
local variable 'track_id' referenced before assignment missing Edna mechta 
Unable to get Uvek kad u nebo pogledam  from Maja Mazić
local variable 'track_id' referenced before

Unable to get Ti ne me poznavaš  from Marija & Viktorija
local variable 'track_id' referenced before assignment missing Ti ne me poznavaš 
Unable to get Coś mnie nosi from Katarzyna Żurawik
local variable 'track_id' referenced before assignment missing Coś mnie nosi
Unable to get Sinnsykt gal forelsket from 2U
local variable 'track_id' referenced before assignment missing Sinnsykt gal forelsket
Unable to get Tobele sunt viața mea from Bubu
local variable 'track_id' referenced before assignment missing Tobele sunt viața mea
Unable to get De vriendschapsband from X!NK
local variable 'track_id' referenced before assignment missing De vriendschapsband
Unable to get My Song for the World from Tom Morley
local variable 'track_id' referenced before assignment missing My Song for the World
Unable to get Stoppa mig from The Honeypies
local variable 'track_id' referenced before assignment missing Stoppa mig
Unable to get Like a Star from Sarah Harrison
local variable 'track_id' referenced before

Unable to get Sweetie Baby from Compass Band
local variable 'track_id' referenced before assignment missing Sweetie Baby
Unable to get Nebo  from Anastasiya Petryk
local variable 'track_id' referenced before assignment missing Nebo 
Unable to get Funky Lemonade from Funkids
local variable 'track_id' referenced before assignment missing Funky Lemonade
Unable to get Toate vor fi from Denis Midone
local variable 'track_id' referenced before assignment missing Toate vor fi
Unable to get Power of a Song from Young Talent Team
local variable 'track_id' referenced before assignment missing Power of a Song
Unable to get Hij is een kei from Klaartje & Nicky
local variable 'track_id' referenced before assignment missing Hij is een kei
Unable to get Birichino from Demis Mirarchi
local variable 'track_id' referenced before assignment missing Birichino
Unable to get En stjerne skal jeg bli from @lek
local variable 'track_id' referenced before assignment missing En stjerne skal jeg bli
Unable to get

In [4]:
train=pd.concat(lista)
train.to_csv("training.csv",index=False)
train

duration  key  loudness  mode    tempo  artist_hotttnesss  \
9    164.78517    0    -3.892     0  173.703               0.03   
15   181.99773    7    -9.342     0   95.001               0.23   
16   179.98036    0    -3.938     1  113.932               0.26   
17   182.60023    2    -7.322     1  113.981               0.17   
20   173.16830    6    -6.834     1  101.021               0.35   
..         ...  ...       ...   ...      ...                ...   
255  168.04000    2    -6.348     0  134.025               0.33   
257  320.30110    8    -3.767     1  130.034               0.33   
258  214.00000    4   -15.249     1  136.034               0.01   
261  142.53333    8    -9.105     1   90.247               0.32   
268  167.45332    6    -4.085     1  112.028               0.33   

     end_of_fade_in  start_of_fade_out  mode_confidence  key_confidence  ...  \
9           0.00000          164.78517            0.487           0.308  ...   
15          0.00000          178.39311            0.873           0.876  ...   
16          0.00000          173.35439            0.636           0.742  ...   
17          0.24989          176.25687            0.687           0.619  ...   
20          0.15116          168.11247            0.768           0.742  ...   
..              ...                ...              ...             ...  ...   
255         0.36594          163.22467            0.427           0.266  ...   
257         0.00000          316.44736            0.513           0.466  ...   
258         0.00000          211.68472            0.387           0.501  ...   
261         0.00000          138.80309            0.602           0.578  ...   
268         0.10730          163.57297            0.556           0.686  ...   

     danceability  energy speechiness  acousticness  instrumentalness  \
9           0.453   0.888      0.0634       0.09830          0.000000   
15          0.565   0.563      0.0296       0.13100          0.000002   
16          0.758   0.647      0.0419       0.43300          0.000000   
17          0.667   0.405      0.0292       0.54700          0.000000   
20          0.611   0.623      0.0367       0.05690          0.000045   
..            ...     ...         ...           ...               ...   
255         0.775   0.683      0.0480       0.07720          0.000000   
257         0.655   0.989      0.0494       0.03720          0.018000   
258         0.839   0.707      0.0666       0.00116          0.918000   
261         0.616   0.893      0.0812       0.07460          0.000078   
268         0.688   0.688      0.0268       0.44500          0.000000   

     liveness  valence  duration_ms  Points       Country  
9      0.4110    0.812       164785   100.0         Malta  
15     0.2530    0.114       181998     NaN       Belarus  
16     0.1720    0.597       179980     NaN        France  
17     0.1920    0.329       182600     NaN       Germany  
20     0.0930    0.428       173168     NaN   Netherlands  
..        ...      ...          ...     ...           ...  
255    0.3530    0.580       168040   103.0   Netherlands  
257    0.5930    0.761       320301    57.0        Sweden  
258    0.1690    0.965       214000   108.0       Georgia  
261    0.0786    0.890       142533   121.0       Denmark  
268    0.3060    0.687       167453    82.0   Netherlands  

[92 rows x 24 columns]

## Ahora importaremos el CSV con la información de los participantes del 2020 (sin puntuación ya que el evento todavía no ha ocurrido, esto es lo que nos gustaría predecir).

In [9]:
df_2020=pd.read_csv("2020.csv")

In [10]:
lista2=[]
for index, row in df_2020.iterrows():
    try:
        dic = get_info(song_name=row["Song"],artist_name=row["Artist"],force=True)
        dic['Country']=row['Country']
        lista2.append(pd.DataFrame(dic,index=[index]))
        
    except Exception as e:
        print(e,f"missing {row['Artist'],row['Song']}")

Unable to get Forever from Karakat Bashanova
Trying Forever by only artist Karakat Bashanova
list index out of range missing ('Karakat Bashanova', 'Forever')
Unable to get Chasing Sunsets from Chanel Monseigneur
Trying Chasing Sunsets by only artist Chanel Monseigneur
list index out of range missing ('Chanel Monseigneur', 'Chasing Sunsets')
Unable to get Ill Be Standing from Ala Tracz
Trying Ill Be Standing by only artist Ala Tracz
Unable to get Heartbeat from Petar Aničić
Trying Heartbeat by only artist Petar Aničić
list index out of range missing ('Petar Aničić', 'Heartbeat')


### Finalmente terminamos con nuestro conjunto del año 2020 enriquecido por las APIs de Spotify (sin la columna de puntuación)

In [11]:
final=pd.concat(lista2)
pd.set_option('display.max_columns', 500)
final.to_csv('final.csv',index=False)
final

duration  key  loudness  mode    tempo  artist_hotttnesss  \
0   181.99773    7    -9.342     0   95.001               0.23   
1   179.98036    0    -3.938     1  113.932               0.26   
2   182.60023    2    -7.322     1  113.981               0.17   
5   173.16830    6    -6.834     1  101.021               0.35   
6   180.54675    0    -2.799     1  122.028               0.31   
7   177.33333    8    -6.671     1  180.020               0.14   
9   167.98611    6    -6.184     0  100.040               0.30   
10  157.90765   11    -7.370     0  153.369               0.19   

    end_of_fade_in  start_of_fade_out  mode_confidence  key_confidence  \
0          0.00000          178.39311            0.873           0.876   
1          0.00000          173.35439            0.636           0.742   
2          0.24989          176.25687            0.687           0.619   
5          0.15116          168.11247            0.768           0.742   
6          0.00000          173.85940            0.490           0.546   
7          0.00000          169.28508            0.311           0.507   
9          2.61805          162.60934            0.379           0.062   
10         0.53991          152.74086            0.655           0.696   

    time_signature  time_signature_confidence  year  popularity  danceability  \
0                4                      0.941  2020          35         0.565   
1                4                      1.000  2020          38         0.758   
2                4                      0.783  2020          28         0.667   
5                4                      1.000  2020          46         0.611   
6                4                      0.899  2020          42         0.523   
7                4                      1.000  2020          26         0.258   
9                4                      1.000  2020          42         0.744   
10               4                      0.973  2020          31         0.359   

    energy  speechiness  acousticness  instrumentalness  liveness  valence  \
0    0.563       0.0296        0.1310          0.000002    0.2530    0.114   
1    0.647       0.0419        0.4330          0.000000    0.1720    0.597   
2    0.405       0.0292        0.5470          0.000000    0.1920    0.329   
5    0.623       0.0367        0.0569          0.000045    0.0930    0.428   
6    0.851       0.0373        0.0148          0.000001    0.2920    0.181   
7    0.499       0.0377        0.4550          0.000000    0.0773    0.428   
9    0.574       0.1670        0.0555          0.000000    0.0817    0.353   
10   0.497       0.0439        0.4360          0.000000    0.0787    0.328   

    duration_ms       Country  
0        181998       Belarus  
1        179980        France  
2        182600       Germany  
5        173168   Netherlands  
6        180547        Poland  
7        177333        Russia  
9        167986         Spain  
10       157908       Ukraine